src/future/shared.rs

```diff
@@ -0,0 +1,229 @@
+//! assert_eq!(6, *shared1.wait().unwrap());
+//! assert_eq!(6, *shared2.wait().unwrap());
+//! ```
+
+use std::mem;
+use std::vec::Vec;
+use std::sync::{Arc, RwLock};
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::ops::Deref;
+
+use {Future, Poll, Async};
+use task::{self, Task};
+use lock::Lock;
+
+
+/// A future that is cloneable and can be polled in multiple threads.
+/// Use Future::shared() method to convert any future into a `Shared` future.
+#[must_use = "futures do nothing unless polled"]
+pub struct Shared<F>
+    where F: Future
+{
+    inner: Arc<Inner<F>>,
+}
+
+struct Inner<F>
+    where F: Future
+{
+    /// The original future.
+    original_future: Lock<F>,
+    /// Indicates whether the result is ready, and the state is `State::Done`.
+    result_ready: AtomicBool,
+    /// The state of the shared future.
+    state: RwLock<State<F::Item, F::Error>>,
+}
+
+/// The state of the shared future. It can be one of the following:
+/// 1. Done - contains the result of the original future.
+/// 2. Waiting - contains the waiting tasks.
+enum State<T, E> {
+    Waiting(Vec<Task>),
+    Done(Result<SharedItem<T>, SharedError<E>>),
+}
+
+impl<F> Shared<F>
+    where F: Future
+{
+    /// Creates a new `Shared` from another future.
+    pub fn new(future: F) -> Self {
+        Shared {
+            inner: Arc::new(Inner {
+                original_future: Lock::new(future),
+                result_ready: AtomicBool::new(false),
+                state: RwLock::new(State::Waiting(vec![])),
+            }),
+        }
+    }
+
+    /// Converts a result as it's stored in `State::Done` into `Poll`.
+    fn result_to_polled_result(result: Result<SharedItem<F::Item>, SharedError<F::Error>>)
+                               -> Result<Async<SharedItem<F::Item>>, SharedError<F::Error>> {
+        match result {
+            Ok(item) => Ok(Async::Ready(item)),
+            Err(error) => Err(error),
+        }
+    }
+
+    /// Clones the result from self.inner.state.
+    /// Assumes state is `State::Done`.
+    fn read_result(&self) -> Result<Async<SharedItem<F::Item>>, SharedError<F::Error>> {
+        match *self.inner.state.read().unwrap() {
+            State::Done(ref result) => Self::result_to_polled_result(result.clone()),
+            State::Waiting(_) => panic!("read_result() was called but State is not Done"),
+        }
+    }
+
+    /// Stores the result in self.inner.state, unparks the waiting tasks,
+    /// and returns the result.
+    fn store_result(&self,
+                    result: Result<SharedItem<F::Item>, SharedError<F::Error>>)
+                    -> Result<Async<SharedItem<F::Item>>, SharedError<F::Error>> {
+        let ref mut state = *self.inner.state.write().unwrap();
+
+        match mem::replace(state, State::Done(result.clone())) {
+            State::Waiting(waiters) => {
+                self.inner.result_ready.store(true, Ordering::Relaxed);
+                for task in waiters {
```

Comment:
No, because otherwise another thread could call read_result() that will panic if the state is State::Waiting